File: binarysearch.io/October 2020/256_remove_duplicates_in_ll.cpp
#include "solution.hpp"
using namespace std;

/**
 * class LLNode {
 *     public:
 *         int val;
 *         LLNode *next;
 * };
 */
class Solution {
    public:
    LLNode* solve(LLNode* node){
        unordered_map<int,int> hashmap;
        LLNode * head = node;
        LLNode * prev;
        while(node){
            if(hashmap.find(node->val) != hashmap.end()){
                prev->next = node->next;
            }else{
                prev = node;
            }
            hashmap[node->val] = 1;
            node = node->next;
        }
        return head;
    }
};

File: binarysearch.io/October 2020/256_remove_duplicates_in_ll.py
# class LLNode:
#     def __init__(self, val, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def solve(self, node):
        duplicates = set()
        head = node 
        prev = None 
        while head:
            if head.val in duplicates:
                prev.next = head.next
            else:
                prev = head 
            duplicates.add(head.val)
            head = head.next
        
        return node 
        
        
